cluster-autoscaler/core/scaledown/planner/planner_test.go

```diff
@@ -0,0 +1,401 @@
+		},
+		{
+			name: "pods from already drained node can schedule elsewhere",
+			nodes: []*apiv1.Node{
+				BuildTestNode("n1", 1000, 10),
+				nodeUndergoingDeletion("n2", 2000, 10),
+			},
+			pods: []*apiv1.Pod{
+				scheduledPod("p1", 500, 1, "n2"),
+				scheduledPod("p2", 500, 1, "n2"),
+			},
+			eligible: []string{"n1"},
+			actuationStatus: &fakeActuationStatus{
+				currentlyDrained: []string{"n2"},
+			},
+			wantUnneeded: []string{},
+		},
+		{
+			name: "pods from already drained node can't schedule elsewhere",
+			nodes: []*apiv1.Node{
+				BuildTestNode("n1", 1000, 10),
+				nodeUndergoingDeletion("n2", 2000, 10),
+			},
+			pods: []*apiv1.Pod{
+				scheduledPod("p1", 500, 1, "n2"),
+				scheduledPod("p2", 500, 1, "n2"),
+				scheduledPod("p3", 500, 1, "n2"),
+			},
+			eligible: []string{"n1"},
+			actuationStatus: &fakeActuationStatus{
+				currentlyDrained: []string{"n2"},
+			},
+			wantUnneeded: []string{},
+			wantErr:      true,
+		},
+		{
+			name: "pods from multiple drained nodes can schedule elsewhere",
+			nodes: []*apiv1.Node{
+				BuildTestNode("n1", 1000, 10),
+				nodeUndergoingDeletion("n2", 2000, 10),
+				BuildTestNode("n3", 1000, 10),
+				nodeUndergoingDeletion("n4", 2000, 10),
+			},
+			pods: []*apiv1.Pod{
+				scheduledPod("p1", 500, 1, "n2"),
+				scheduledPod("p2", 500, 1, "n2"),
+				scheduledPod("p4", 500, 1, "n4"),
+				scheduledPod("p5", 500, 1, "n4"),
+			},
+			eligible: []string{"n1", "n3"},
+			actuationStatus: &fakeActuationStatus{
+				currentlyDrained: []string{"n2", "n4"},
+			},
+			wantUnneeded: []string{},
+		},
+		{
+			name: "pods from multiple drained nodes can't schedule elsewhere",
+			nodes: []*apiv1.Node{
+				BuildTestNode("n1", 1000, 10),
+				nodeUndergoingDeletion("n2", 2000, 10),
+				BuildTestNode("n3", 1000, 10),
+				nodeUndergoingDeletion("n4", 2000, 10),
+			},
+			pods: []*apiv1.Pod{
+				scheduledPod("p1", 500, 1, "n2"),
+				scheduledPod("p2", 500, 1, "n2"),
+				scheduledPod("p3", 500, 1, "n2"),
+				scheduledPod("p4", 500, 1, "n4"),
+				scheduledPod("p5", 500, 1, "n4"),
+			},
+			eligible: []string{"n1", "n3"},
+			actuationStatus: &fakeActuationStatus{
+				currentlyDrained: []string{"n2", "n4"},
+			},
+			wantUnneeded: []string{},
+			wantErr:      true,
+		},
+		{
+			name: "multiple drained nodes but new candidates found",
+			nodes: []*apiv1.Node{
+				BuildTestNode("n1", 2000, 10),
+				nodeUndergoingDeletion("n2", 2000, 10),
+				BuildTestNode("n3", 2000, 10),
+				nodeUndergoingDeletion("n4", 2000, 10),
+				BuildTestNode("n5", 2000, 10),
+			},
+			pods: []*apiv1.Pod{
+				scheduledPod("p1", 400, 1, "n1"),
+				scheduledPod("p2", 400, 1, "n2"),
+				scheduledPod("p3", 400, 1, "n3"),
+				scheduledPod("p4", 400, 1, "n4"),
+				scheduledPod("p5", 400, 1, "n5"),
+			},
+			eligible: []string{"n1", "n3", "n5"},
+			actuationStatus: &fakeActuationStatus{
+				currentlyDrained: []string{"n2", "n4"},
+			},
+			wantUnneeded: []string{"n1", "n3"},
```

Comment:
How do we know `n5` is the one needed, and not `n1`, or `n3`? Can't `p2` and `p4` be assigned e.g. to `n1`? The same applies to some other test cases.

Reply:
The simulation is deterministic: it evaluates the nodes in the order in which they are provided. This is an intended feature, added a comment to `Planner`'s public interface to make it clear.